### DISCUSSION TITLE
How to match LaTeX commands with ltex.hiddenFalsePositives?

### DISCUSSION THREAD

**Discussion by thbtppl** [valentjn/vscode-ltex#202] [Q&A] [answered]
Thanks for that great VSC extension.
I find the documentation ambiguous in how to set that setting explicitly `ltex.hiddenFalsePositive` in the settings file.

If I do

`"ltex.enabledRules": {"en-GB": ["{"rule": "UPPERCASE_SENTENCE_START", "sentence": "\\\\gls"}"]}`

I get `Expected comma`
What's the correct syntax?

**Comment by valentjn** [comment]
- Use the correct key, i.e., `ltex.hiddenFalsePositives` instead of `ltex.enabledRules`.
- Escape quotation marks inside JSON strings.
- JSON objects need to be closed with a `}`.
- Use the `Hide false positive` quickfix to see an example of the syntax, and modify it to your liking.

Try the following: `"ltex.hiddenFalsePositives": {"en-GB": ["{\"rule\": \"UPPERCASE_SENTENCE_START\", \"sentence\": \"\\\\gls\"}"]}`

**Comment by thbtppl** [comment, marked as answer]
```
    "ltex.hiddenFalsePositives": {
        "en-GB": [
            "{\"rule\": \"UPPERCASE_SENTENCE_START\", \"sentence\": \"\\\\gls\"}",
            "{\"rule\": \"UPPERCASE_SENTENCE_START\", \"sentence\": \"\\\\lettrine\"}"
        ]
    }
```
Putting this results in an internal error of ltex when I want to check the document.

Alternatively, looking at the syntax of the generated file with the quick fix:

`{"rule":"UPPERCASE_SENTENCE_START","sentence":"^\\Qxxxxxxxxxxxxxxxxxx\\E$"}`

Are the `^\\Q` and `\\E$` necessary if I were to define my own patterns? 

Basically I would like to disable reporting of these false positives whenever I have a \gls or \lettrine in ANY sentence of the document.

**Reply by valentjn** [reply]
> Basically I would like to disable reporting of these false positives whenever I have a \gls or \lettrine in ANY sentence of the document.

Ah, I see. That's not supported. The documentation is not perfect here, I'll fix that.

When you check a L<sup>A</sup>T<sub>E</sub>X document with LT<sub>E</sub>X, LT<sub>E</sub>X converts the L<sup>A</sup>T<sub>E</sub>X code to "plain text." That's basically the text parts of the code stripped by all the L<sup>A</sup>T<sub>E</sub>X commands and environments that might be in there. The plain text is passed to LanguageTool (LT), which splits the plain text into sentences and returns a list of rule matches. A rule match contains the position of the match in the original code, which is then used by VS Code to underline the error. Along with that, a rule match also contains the sentence, in which the rule match is, as an [`AnalyzedSentence`](https://github.com/languagetool-org/languagetool/blob/a31350879aa38735d4416116b96df3ceea392a9b/languagetool-core/src/main/java/org/languagetool/AnalyzedSentence.java). The problem is that the sentence is in the plain text, not in the original code. I don't see a straightforward way to get the sentence in the original code (start and end position in either original code or plain text would suffice). I'm happy for suggestions, but otherwise I think this is an upstream issue in LT.

So in short, the problem is that false positive "sentences" are plain text sentences in which all the L<sup>A</sup>T<sub>E</sub>X code has been stripped away.

As a workaround, I suggest
- to add `"\\gls{}": "dummy"` to [`ltex.latex.commands`](https://valentjn.github.io/vscode-ltex/docs/settings.html#ltexlatexcommands) (dummy words are capitalized), and
- to hide `\lettrine` false positives individually with the quick fix (`\lettrine` is usually only used at the beginning of a chapter, so there are probably not that many false positives). I'm not sure about your use case though &mdash; LT<sub>E</sub>X already replaces `\lettrine{}{}` and `\lettrine[]{}{}` with capitalized dummy words by default.

The reason for the "internal error" is that the statement about escaping a backslash with four backslashes is about escaping backslashes in the regular expression. If you want to match a backslash (e.g., the one in `\gls`) in regular expressions, you'll have to use two backslashes, each of which has to be escaped twice due to the double JSON parsing, so you need eight backslashes in this case. This will also be added to the docs.

**Reply by valentjn** [reply]
[The docs](https://valentjn.github.io/vscode-ltex/docs/advanced-usage.html#hiding-false-positives-with-regular-expressions) have now been edited to reflect what I wrote. Thanks for the suggestion.
